poetry_to_pre_commit/common.py

```diff
@@ -23,10 +24,33 @@ def pre_commit_config_roundtrip(
         yaml.dump(config, path)
 
 
-@dataclasses.dataclass
+@dataclasses.dataclass(frozen=True)
 class PoetryPackage:
     name: str
-    version: str
+    version: str = "*"
+    extras: frozenset[str] = dataclasses.field(default_factory=frozenset)
+
+    def __str__(self) -> str:
+        # Since the object is frozen, we can cache the result. But applying
+        # `lru_cache` on `__str__` makes trouble, so we use a cached property.
+        return self._str
+
+    def __lt__(self, value: object) -> bool:
+        return str(self) < str(value)
+
+    def __eq__(self, value: object) -> bool:
+        return str(self) == str(value)
+
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+    @cached_property
+    def _str(self) -> str:
+        if self.extras:
+            extras = ",".join(sorted(self.extras))
+            return f"{self.name}[{extras}]=={self.version}"
+        else:
+            return f"{self.name}=={self.version}"
```

Comment:
Hm? I'm not sure I understand why we need that ? dataclasses already provide this, no?

Reply:
`__hash__` is needed because `extras` is unhashable.
 `__lt__` allows to sort a set of `PoetryPackage`.
And if you ovewrite `__lt__`, it makes sense to overwrite `__eq__` too :-)

I added comments next to `__lt__` and `__hash__`.

Reply:
Hm ? 
```python
>>> {frozenset(["a", "b"])}
{frozenset({'a', 'b'})}

>>> import dataclasses
>>> @dataclasses.dataclass(frozen=True)
... class X:
...     extras: frozenset[str] = dataclasses.field(default_factory=frozenset)
... 
>>> {X()}
{X(extras=frozenset())}

```
Frozensets of strings are hashable, no ?

> `__lt__` allows to sort a set of `PoetryPackage`
You can use `dataclass(order=True)` for this

Reply:
> You can use dataclass(order=True) for this

Nice! No more dunder methods in `common.PoetryPackage` :-)